### DISCUSSION TITLE
Template for new jails

### DISCUSSION THREAD

**Discussion by spusuf** [Jip-Hop/jailmaker#21] [General]
On every jail I've made I found myself using the same few scripts, and it would be nice to have these all done in one "base" or "template" jail to then copy to new installs for an outdated but ready to run install without even having to open the jail's shell. 

For now I'd simply have a jail automatically install curl, openssh, and docker, but it would be nice to not start from scratch for different needs.

My (basic) commands on every install currently are:
apt update, install openssh, install curl, curl get docker convenience script, run docker install script.

**Comment by Jip-Hop** [comment]
What you could already do is:

1. Make a shell script with those template commands and store it somewhere on the NAS.
2. Make this script or directory available to a new jail using bind mount (possibly read only) additional flags during the guided steps from `./jlmkr.py create myjail`.
3.  When the jail has been created and started, you can run your template script using this oneliner: `systemd-run --machine myjail --quiet --pipe --wait --collect --service-type=exec bash /path/to/your/template/script.sh`.

Alternatively you can create one 'template' jail manually and then just copy this jail to a new directory. Say your jail is here `/mnt/mypool/jailmaker/jails/mytemplatejail`. You can then run `cp -rp /mnt/mypool/jailmaker/jails/mytemplatejail /mnt/mypool/jailmaker/jails/derivativejail` and you'll have another jail based on the template. It would be good though to ensure the template jail is not running when you're doing the copy.

**Comment by Jip-Hop** [comment]
By the way I don't think you need openssh and curl in each jail, as curl is already available on the host (SCALE) and SCALE also has an SSH server. So if you want to open a shell in the jail just ssh into SCALE and run `machinectl shell myjail`. And if you want to curl a file you can do so from the host and put it in a directory which is available to the jail.

**Reply by spusuf** [reply]
Curl is for the Docker install script, having it on the host did not allow me to use it in the jail.

Openssh will be for me to be able to SSH into a particular service/app's jail instead of the TrueNAS host (which I want to disable SSH for security)

**Comment by spusuf** [comment]
Copying a containers directory wouldnt change the hostname and other misc settings, so I have settled by creating a new jail then running this command in the TrueNAS host, manually replacing `myjail` with the actual container name, in order to get my basic functionality for every jail:

`jlmkr exec myjail bash -c 'sudo apt update; sudo apt install nano -y; sudo apt install git -y; sudo apt install curl -y; curl -fsSL https://get.docker.com -o get-docker.sh; sudo sh get-docker.sh'`

**Reply by Jip-Hop** [reply]
Nice :) Do you really need sudo by the way?

**Reply by spusuf** [reply]
Is it strictly necessary? Probably not. Do all the guides I follow recommend it? Yes.

**Reply by Jip-Hop** [reply]
I think exec already runs the commands as root so then it's not needed. The guides assume you're not already running as root. Just a detail :)

**Comment by PackElend** [comment]
> `jlmkr exec myjail bash -c 'sudo apt update; sudo apt install nano -y; sudo apt install git -y; sudo apt install curl -y; curl -fsSL https://get.docker.com -o get-docker.sh; sudo sh get-docker.sh'`

thanks a lot for this, I was not able to get the get-docker.sh running (using [filebrowser](https://github.com/filebrowser/filebrowser) to create the file)).
I only needed to delete the 'sudo'.

**Comment by Jip-Hop** [comment]
I'm locking this discussion as I've opened dedicated issues to track feature requests. Please see [Template for new jails](https://github.com/Jip-Hop/jailmaker/issues/84).
